File: app/src/main/java/com/proxy/shadowsocksr/adapter/AppsAdapter.java
package com.proxy.shadowsocksr.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import com.proxy.shadowsocksr.R;
import com.proxy.shadowsocksr.adapter.items.AppItem;

import java.util.List;

public final class AppsAdapter extends RecyclerView.Adapter<AppsAdapter.ViewHolder>
{
    private List<AppItem> appLst;
    private OnItemClickListener onItemClickListener;

    public AppsAdapter(List<AppItem> appLst)
    {
        this.appLst = appLst;
    }

    @Override public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType)
    {
        View v = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.list_app_item, parent, false);
        return new ViewHolder(v);
    }

    @Override public void onBindViewHolder(ViewHolder vh, int position)
    {
        AppItem ai = appLst.get(position);
        vh.tv.setText(ai.name);
        vh.iv.setImageDrawable(ai.icon);
        vh.cb.setChecked(ai.checked);
    }

    @Override public int getItemCount()
    {
        return appLst.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener
    {
        public ImageView iv;
        public TextView tv;
        public CheckBox cb;

        public ViewHolder(View v)
        {
            super(v);
            iv = (ImageView) v.findViewById(R.id.iv);
            tv = (TextView) v.findViewById(R.id.tv_name);
            cb = (CheckBox) v.findViewById(R.id.cb);
            v.setOnClickListener(this);
        }

        @Override public void onClick(View v)
        {
            if (onItemClickListener != null)
            {
                appLst.get(getAdapterPosition()).checked = !cb.isChecked();
                cb.setChecked(!cb.isChecked());
                onItemClickListener.onItemClick(v, getAdapterPosition());
            }
        }
    }

    public interface OnItemClickListener
    {
        void onItemClick(View v, int pos);
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener)
    {
        this.onItemClickListener = onItemClickListener;
    }
}

File: jvmVer/src/main/java/com/proxy/shadowsocksr/adapter/AppsAdapter.kt
package com.proxy.shadowsocksr.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.TextView

import com.proxy.shadowsocksr.R
import com.proxy.shadowsocksr.adapter.items.AppItem

class AppsAdapter(private val appLst: List<AppItem>) : RecyclerView.Adapter<AppsAdapter.ViewHolder>()
{
    var onItemClickListener: OnItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder
    {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.list_app_item, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(vh: ViewHolder, position: Int)
    {
        val ai = appLst[position]
        vh.tv.text = ai.name
        vh.iv.setImageDrawable(ai.icon)
        vh.cb.isChecked = ai.checked
    }

    override fun getItemCount(): Int
    {
        return appLst.size
    }

    public inner class ViewHolder(v: View) : RecyclerView.ViewHolder(v), View.OnClickListener
    {
        var iv: ImageView
        var tv: TextView
        var cb: CheckBox

        init
        {
            iv = v.findViewById(R.id.iv) as ImageView
            tv = v.findViewById(R.id.tv_name) as TextView
            cb = v.findViewById(R.id.cb) as CheckBox
            v.setOnClickListener(this)
        }

        override fun onClick(v: View)
        {
            if (onItemClickListener != null)
            {
                cb.isChecked = !cb.isChecked
                appLst[adapterPosition].checked = cb.isChecked
                onItemClickListener!!.onItemClick(v, adapterPosition)
            }
        }
    }

    interface OnItemClickListener
    {
        fun onItemClick(v: View, pos: Int)
    }
}
